### DISCUSSION TITLE
Debugging non-interactive sessions

### DISCUSSION THREAD

**Discussion by bfv** [vscode-abl/vscode-abl#117] [Q&A]
I'm investigation the possibility to use vscode devcontainers (see: https://github.com/bfv/openedge-devcontainer) and what I encounter is that I cannot use the debugger in the container. This is probably due to the start of an interactive session (which display a message box), but that doesn't compute in a container. Is the extension in its current form able to support such a use case?

**Comment by gquerret** [comment]
How do you start the client session ?

**Comment by bfv** [comment]
I use `launch.json`:
```
{
    // Use IntelliSense to learn about possible attributes.
    // Hover to view descriptions of existing attributes.
    // For more information, visit: https://go.microsoft.com/fwlink/?linkid=830387
    "version": "0.2.0",
    "configurations": [
        {
            "name": "Debug program / Interactive TTY",
            "type": "abl",
            "request": "launch",
            "program": "${file}",
            "cwd": "${workspaceFolder}",
            "graphicalMode": false
        },
    ]
}
```
Although it says "Interactive TTY", I wouldn't know how to make it non-interactive.

**Comment by kenherring** [comment]
Following, I'm interested as well.

My current workaround is:

* use the `attach` method.
* Start a `-b` session in the terminal using a **preLaunchTask** 
  * the task uses a background type
  * code waits for the debugger to connect before starting the main block
* The debugger launches and connects to the the started terminal process

Took a while to get that working so something more native would be awesome.

Another option might be [compound launch configuarions](https://code.visualstudio.com/docs/editor/debugging#_compound-launch-configurations) but I haven't had a chance to try it out yet.

**Comment by PeterJudgeZA** [comment]
I ended up putting a pause block to my .P and attach manually ... not great but it works OK. 

```
define variable i as integer no-undo init 0.        //set to 20 secs for debugging

do while i > 0:
    message "waiting " i " secs for debugger attach ...".
    pause 1 no-message.
    i -= 1.
end.
```

**Comment by gquerret** [comment]
I don't think there's a way to have a batch session wait for the debugger to be attached, except using Peter's trick. It may be worth opening a ticket with TS. @bfv Do you have time for that ?

**Comment by kenherring** [comment]
We use something very similar to @PeterJudgeZA, but add in `DEBUGGER:VISIBLE` to see if it's connected.  Works great ahead of unit tests, among other things.

Something along these lines:

```abl
define variable cnt as integer no-undo.

waitBlock:
for cnt = 0 to 30:
    message "waited " + string(cnt) " seconds for debugger attach ...".
    pause 1 no-message. // Click continue or press f5 to continue debugging
    if debugger:visible then
       leave waitBlock.
end.
```

**Reply by PeterJudgeZA** [reply]
Unit tests is also my use-case here ...

**Comment by bfv** [comment]
Thanks for the answers so far. Today I have a day off, I'll give this a shot tomorrow. The fact that everything runs in a container seems to complicate matters, not sure why though because debugging is (should be) possible in devcontainers. Btw, the ultimate goal here is to be able to provide an entire (backend) development environment by just pointing to a GitHub repository. @gquerret I'm interested if you see any pitfalls in my approach.

**Reply by kenherring** [reply]
I've set up a fully functional devcontainer, including debugging.  Can confirm it works with a lot of trial and error.  Haven't had a chance to try using it for codespaces yet, but someday I'll find the time.

**Reply by gquerret** [reply]
I currently don't see any problems, and I'm interested in seeing a functional environment in GH codespaces. I can help with issues in this environment, as long as I can reproduce them, so make sure I have access to the repo, and that I can easily switch to my own Docker images.

**Reply by kenherring** [reply]
@bfv Not sure if it helps with anything but I have some code from the pug challenge referencing dev containers: https://github.com/kenherring/pug-challenge/tree/main/.devcontainer

Can't guarantee how much functionality actually works since I haven't played with it in a while.  Will likely continue to update this repo whenever I experiment with new things, but no immediate plans. Cheers

**Reply by gquerret** [reply]
Is it possible to have multiple containers in this configuration ? For example, one container for the language server, one for the database (or maybe one for each database), and one for PASOE.

**Reply by bfv** [reply]
@gquerret Yes, that's the whole point of pursuing that. Once the "heavy lifting" is done, a developer should be able to download the git repo and tart developing. A few ceveats:
- vscode should be installed locally with the devcontainer plugin
- Docker (Desktop) should be installed
- The license should be somewhere locally

In theory this make it possible to develop on a Mac, or to develop on any machine without installing OpenEdge altogether.

**Reply by kenherring** [reply]
> Is it possible to have multiple containers in this configuration ? For example, one container for the language server, one for the database (or maybe one for each database), and one for PASOE.

Depending on what we're working on we use [COMPOSE_PROFILES](https://docs.docker.com/compose/profiles/#start-specific-profiles) to spin up a wide variety of containers for local development.  This includes db, pasoe, batch jobs, repl, etc.  In the past we once had 10 database containers (1 each), but have since moved to building a single container with 10 databases.

**Reply by gquerret** [reply]
Single container for all databases is probably easier to manage from an administration perspective. But then you have an extra container per PASOE instance ?
